Issue #46 - Fetch analysis logs / health check

LukaDeka opened the pull request:
## What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
* Feature


## What is the current behavior?** (You can also link to an open issue here)
* Nothing 


## What is the new behavior (if this is a feature change)? If possible add a screenshot.

* Added a scheduled task for the orchestrator that iterates over every busy worker and:
1) zips `settings.WORKER_EMBA_LOGS`
2) downloads them on the local machine
3) unpacks them to the correct destination.


* Added another scheduled task that monitors the health of the workers:
1) It checks if the docker container is still running on the worker
2) if not, it sets the worker as "free" and
3) performs a `worker_soft_reset`.
4) This task can be easily extended. 


* Refactored `workers.tasks::start_analysis`:
1) made the start of the analysis on the worker non-blocking. Otherwise it's really strange.
2) added `LogReader` so the analysis is tracked in the UI.


* Refactored `workers.views::worker_soft_reset`:
1) split the function into a helper function that can be called directly (without a view, i.e., a `request` arg)
2) fixed `config.user != user` assertion logic (as pointed out by Clemens)


* Fixed syntax/logical errors in `uploader.boundedexecutor::unzip_log`:
even though I'm not directly calling this function in my pull request, I discovered errors when testing it out before. I'm pretty sure the function is broken.


* Fixed syntax errors in `workers.tasks::_pase_deb_list`


## Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
--- The refactorings make sure the functions are backwards-compatible.


##Other information
* I had to make a lot of design choices, also with the endpoint. A lot can be improved and is done in a subpar way currently. I left quite a few TODOs and comments.

* A lot will probably need to be changed in this pull request before it's ready

ClProsser commented:
Assigned myself as reviewer as discussed in the team meeting

LukaDeka commented on embark/workers/views.py:
```diff
@@ -295,36 +296,126 @@
     except (Worker.DoesNotExist, Configuration.DoesNotExist):
         return JsonResponse({'status': 'error', 'message': 'Worker or configuration not found.'})
+    except Exception as exception:
+        return JsonResponse({'status': 'error', 'message': f'Unexpected exception: {exception}'})
```
Fixed.

LukaDeka commented on embark/workers/views.py:
```diff
@@ -421,3 +517,95 @@ def safe_redirect(request, default):
     return HttpResponseRedirect(referer)
+
+
+def update_system_info(configuration, worker):
```
I did not add this function... It probably accidentally got pasted in here during a merge

ClProsser reviewed (changes requested):
Many changes have to be done to merge this into `main`. I'll give some detailed feedback below.

Things which are implemented, which are not needed or were not part of the issue (especially, as you argued that it was much work):
- Enable/disable log syncing: I don't see any benefit in not syncing logs. We want partial results, we want to retrieve the result from the worker, thus log syncing is somehow always needed. 
- Refactorings e.g. in the `BoundedExecutor` class: Some refactorings don't preserve the original behavior, thus I don't feel good in accepting them
- The queue endpoint on the worker: With this, you assume that EMBArk (not EMBA) is installed on the worker, as otherwise the endpoint would not be available. While this is not intended by Siemens (why should a worker run a webapp?), this is orthogonal to everything we did on the worker so far (e.g. the installer scripts don't setup EMBArk)

In addition, your implementation is not yet compatible with the existing worker functionality (EMBArk does not exist on the worker, defined file paths (e.g. in the settings) are not yet considered)

As already discussed in the meetings and on discord, I think the best way to fulfill the acceptance criteria is doing it like this:
1. The analysis is started using a celery task. This already specifies where the log files are. [This functionality is already present in `main`, [click here](https://github.com/amosproj/amos2025ss01-embark/blob/535c315e1a7dbc83d3110d89fd6de5df550a880d/embark/workers/tasks.py#L130) ]
2. Start a celery task which has some kind of `while(analysis running)` loop. This loop could look like this: 
2.1 copy log files using ssh/sftp to the EMBArk host
2.2 sleep e.g. 10 seconds (to reduce the CPU load) 
2.3 Evaluate if the analysis is still running
Once the analysis is done, a cleanup should be started
3. Start a celery task which starts the `Logreader`, similar to how the Executor works (https://github.com/amosproj/amos2025ss01-embark/blob/535c315e1a7dbc83d3110d89fd6de5df550a880d/embark/uploader/executor.py#L54C32-L54C41)

I still think that this is the best way to implement this. In any case, it is simpler and requires less code. You might want to recycle your existing implementation where possible, but I think it would be good to stick to this workflow.

Feel free to message me anytime you have a question regarding this.

LukaDeka commented on embark/uploader/boundedexecutor.py:
```diff
@@ -297,48 +291,24 @@ def unzip_log(cls, analysis_id, file_loc):
-        except builtins.Exception as exce:
-            logger.error("Unzipping failed: %s", exce)
+                logs_dir = Path(f"{settings.EMBA_LOG_ROOT}/{analysis_id}/")
+                zip_.extractall(path=logs_dir)
```
`zip_check()` failed every time I tested it.

I couldn't understand what it was for, it looks to me like bad code.

The function checks the existence of certain log files and "accepts" the zip file, if these hard-coded files are in it. It also states in a comment that the functionality of a certain file is broken.

I overall struggled understanding the benefit of the check so I removed it.

LukaDeka replied:
Deleted line `318` also has a syntax error since os.walk returns three arguments. If this is executed, it throws an error. Unclear.
```py
for root_, dir_ in os.walk(Path(f"{settings.EMBA_LOG_ROOT}/{analysis_id}/tmp/")):
```

LukaDeka commented on embark/uploader/models.py:
`onexc` failed a linter check so I refactored it.

LukaDeka commented:
Thank you for your feedback!

I refactored most of the changes from the first draft. The flow should be much clearer now.

I left some changes in `BoundedExecutor` since the syntax/semantics was incorrect (i.e., `os.walk` returning three args and not two). I'm pretty sure that part of the program is broken since the part I changed never gets executed.

I had a few other linter errors which I addressed as well.

Other than that, I look forward to your feedback! Thanks again.

ClProsser commented on embark/workers/tasks.py:
```diff
@@ -146,14 +168,155 @@ def update_worker_info():
+    logger.debug("Worker health-check complete.")
+
+
+def check_with_emba_log(worker) -> Worker.AnalysisStatus:
```
I'd not return `Worker.AnalysisStatus`.
I'd rename it to `is_emba_container_running` (or similar), and return `True` if this is the case, otherwise `False`. 
All checks regarding the `emba.log` are performed in the `LogReader`

ClProsser commented on embark/workers/tasks.py:
```diff
@@ -146,14 +168,155 @@ def update_worker_info():
+            logger.info("[Worker %s] EMBA container is no longer running.", worker.id)
+            return Worker.AnalysisStatus.FREE
+
+        path = f"{settings.WORKER_EMBA_LOGS}emba.log"
+        cmd = f"tail -n 10 {path}"
+        log_output = exec_blocking_ssh(client, cmd)
+        if "Test ended on" in log_output:
+            logger.info("[Worker %s] Analysis finished: %s", worker.id, log_output)
+            return Worker.AnalysisStatus.FREE
```
This is already handled in the `LogReader`

ClProsser commented on embark/workers/tasks.py:
```diff
@@ -146,14 +168,155 @@ def update_worker_info():
+             Worker.AnalysisStatus.RUNNING
+    """
+    try:
+        client = worker.ssh_connect()
```
The `client` is not closed

LukaDeka replied:
Good catch!

ClProsser commented on embark/workers/tasks.py:
```diff
@@ -163,9 +326,6 @@ def start_analysis(worker_id, emba_cmd: str, src_path: str, target_path: str):
 
     client = worker.ssh_connect()
 
-    exec_blocking_ssh(client, f"sudo rm -rf {settings.WORKER_FIRMWARE_DIR}")
-    exec_blocking_ssh(client, f"sudo mkdir -p {settings.WORKER_FIRMWARE_DIR}")
-
```
Why should we not do this here? How to ensure that this folder exists? After installation, this folder does not exist. Similar to updates

LukaDeka replied:
Here I assumed the workers were properly reset and setup if `start_analysis` is called. Probably by another function, since the `settings.WORKER_LOGS_DIR` is not being deleted.

This logic should be part of the cleanup and not of starting the analysis (single-responsibility principle).

I'll revert this change and add a TODO

ClProsser commented on embark/workers/views.py:
```diff
@@ -352,34 +353,47 @@ def worker_soft_reset(request, worker_id, configuration_id=None):
+        else:
+            configuration = worker.configurations.filter(user=user).first()
+
+        if configuration.user != user:
+            return JsonResponse({'status': 'error', 'message': 'You are not allowed to access this worker.'})
+
```
This check is only needed in the then-case.
Furthermore, if no configuration is found in the else case, an exception occurs as `first()` returns `None`

LukaDeka replied:
I didn't add this check, the control flow is identical to what's in `main` right now. I just refactored `worker_soft_reset` and added the helper function since I needed to call it directly (without a `request` param).

I moved it inside the `if True:` check, let me know if you find this acceptable, otherwise I'll remote it entirely.

ClProsser commented on embark/embark/settings/deploy.py:
```diff
@@ -315,6 +315,7 @@
 # Worker setup
 WORKER_FILES_PATH = os.path.join(BASE_DIR.parent, "WORKER_FILES")
 WORKER_SETUP_LOGS = os.path.join(WORKER_FILES_PATH, "logs/worker_setup_{timestamp}.log")
+WORKER_ROOT = "/root/"
```
If you actually want to introduce this variable, please use it everywhere where `/root` is used currently. 
I'd not do that if I where you and just remove `WORKER_ROOT`

LukaDeka replied:
Should I hard-code it instead?

LukaDeka replied:
I'll do that for now.

ClProsser commented on embark/embark/settings/dev.py:
```diff
@@ -282,6 +282,7 @@
 # Worker setup
 WORKER_FILES_PATH = os.path.join(BASE_DIR.parent, "WORKER_FILES")
 WORKER_SETUP_LOGS = os.path.join(WORKER_FILES_PATH, "logs/worker_setup_{timestamp}.log")
+WORKER_ROOT = "/root/"
```
If you actually want to introduce this variable, please use it everywhere where `/root` is used currently. 
I'd not do that if I where you and just remove `WORKER_ROOT`

ClProsser commented on embark/workers/tasks.py:
```diff
@@ -20,16 +26,30 @@ def create_periodic_tasks(**kwargs):
+        name='Monitor running workers',
+        task='workers.tasks.monitor_workers',
+    )
+    PeriodicTask.objects.get_or_create(
+        interval=schedule_10m,
+        name='Fetch worker analysis logs',
+        task='workers.tasks.fetch_running_analysis_logs',
+    )
```
For better UX, this should happen way more often (e.g. every 15-30 seconds). What happens if the task is not yet finished and thus overlap? With enough workers, this could happen even at 10 minutes (even if it is unlikely)

LukaDeka replied:
As discussed in PMs, the time of the schedule can be changed arbitrarily, depending on what the use-case is. An additional setting for the orchestrator can also be added simply.

If the task is scheduled every `n` minutes and the average task takes more than `n` minutes, the tasks will queue up. I don't see this as a problem since in that case, the results will be fetched constantly on a "best-effort" basis. It'll essentially become a `while True` loop.

If a lot of tasks queue up and the analyses finish, the tasks will just execute an iterator with zero elements, i.e., very quickly terminate (since it gets the value dynamically from `Orchestrator.get_busy_workers()`.

Again, as discussed in PMs, a `while True` loop has some benefits over a `PeriodicTask`, though I'm unsure of the use-case for now.

LukaDeka commented on embark/uploader/models.py:
```diff
@@ -158,10 +158,13 @@ def delete_fw_pre_delete_post(sender, instance, **kwargs):
-        shutil.rmtree(instance.get_abs_folder_path(), ignore_errors=False, onexc=logger.error("Error when trying to delete %s", instance.get_abs_folder_path()))
-    else:
-        logger.error("No related FW found for delete request: %s", str(sender))
+    try:
+        if sender.file:
+            shutil.rmtree(instance.get_abs_folder_path(), ignore_errors=False)
+        else:
+            logger.error("No related FW found for delete request: %s", str(sender))
+    except Exception as exception:
+        logger.error("Error when trying to delete %s: %s", instance.get_abs_folder_path(), exception)
```
It was failing a linter check, so I rewrote it.

ClProsser replied:
How to you run the linter? Do you use a different python version than the GH workflows? Otherwise the GH workflows would fail

LukaDeka replied:
`python3 --version` returns `Python 3.10.12`, though this is the system python.

I run the linter using `./dev-tools/check_project.sh --fast`, without `pipenv shell`.

The python version on the GH workflow should be `3.12`

I'm not sure how I can update this.

This is the exact linter error:
```
Run pylint on ./uploader/models.py:

************* Module uploader.models
uploader/models.py:162:8: E1123: Unexpected keyword argument 'onexc' in function call (unexpected-keyword-arg)
```

ClProsser replied:
Python 3.10 should not be used anymore. Siemens wants to move away from it (see https://github.com/e-m-b-a/embark/pull/248/files/275846a4ddd2dbb238e356fd21904cdc81d7ba7f#r2107716592). I don't know whether they like the try/except approach: At least they disliked the onerror approach (which works, but is deprecated). 

You can keep this as is, but I recommend you to change your python version. 
> I'm not sure how I can update this.
You can do this by instructing `pipenv` to use a different python version. 
```bash
  MYSQLCLIENT_LDFLAGS='-L/usr/mysql/lib -lmysqlclient -lssl -lcrypto -lresolv' MYSQLCLIENT_CFLAGS='-I/usr/include/mysql/' PIPENV_VENV_IN_PROJECT=1 pipenv install --dev --python 3.12
```

LukaDeka replied:
Unfortunately, this just updated `Pipenv.lock` and my python version stays the same. Even inside `pipenv shell`.

I reverted the changes, so I'm leaving this conversation as resolved. I'll have to make sure the linter doesn't unnecessarily fail next time

LukaDeka commented on embark/workers/tasks.py:
```diff
@@ -146,14 +168,155 @@ def update_worker_info():
+    logger.debug("Busy worker health check is running...")
+
+    orchestrator = get_orchestrator()
+    busy_workers = list(orchestrator.get_busy_workers().values())
+    for worker in busy_workers:
```
Here, I had the problem of "array being changed while iteration". `list(...)` just makes sure, a new array is created, so if a worker gets removed from `busy_workers`, it won't cause a crash.

Let me know if I overlooked something, but I had the issue while testing and `list(...)` fixed it.

LukaDeka reviewed (commented):
I left replies to the comments.

LukaDeka commented on embark.service:
Can you please tell me what the change here is? I can not for the life of me figure it out.

LukaDeka replied:
I thought so too, but the current version does not have a newline.

I thought I removed a newline, but that wasn't it either...

ClProsser commented:
Looks like you have to initialize the `emba.log` until the first sync occurs. Otherwise an error like this might occur:
```txt
2025-06-22 22:36:47,984: ERROR/ForkPoolWorker-2] inotify_event error in /home/root/embark/emba_logs/b5debdc2-d6ba-4f1f-8fc3-9ac295a75247/emba_logs/emba.log:[Errno 2] No such file or directory
```

ClProsser reviewed (changes requested):
I tested it, however, I could not get it to work. See my comments.

In addition, there is a merge conflict with `main`.

LukaDeka commented on embark/workers/tasks.py:
```diff
@@ -146,14 +167,141 @@ def update_worker_info():
+
+        logger.info("[Worker %s] Zipping logs on remote...", worker.id)
+
+        zip_cmd = "cd /root && \
+                   7z u -t7z -y emba_logs.zip emba_logs/ -uq3"
```
I tested `zip` but that wasn't installed on my machine. What Ubuntu version are you using? Because `7zip` should be installed by default on Ubuntu-22.04

LukaDeka replied:
As discussed in PMs, `tar` doesn't have the functionality to "sync" between the already existing archive and the OS files. See the `--filesync` flag for `zip` for exactly what I mean.

With `tar`, the archive would need to be created from scratch every time to ensure no temp files get left over (the update flag doesn't cover this).

Because of this, adding `7z`, which is already installed on the default Ubuntu 22.04 server edition, to the dependencies makes more sense.

ashiven commented:
I left some comments regarding the orchestrator yesterday but it seems like they were not visible until now?

ashiven reviewed (changes requested):
I hope its okay for me to leave a review as well but I just wanted to point out some things regarding the orchestrator that we should look into

ashiven commented on embark/workers/tasks.py:
```diff
@@ -146,14 +167,141 @@ def update_worker_info():
+
+    logger.debug("Busy worker health check is running...")
+
+    orchestrator = get_orchestrator()
```
I never worked with distributed locks to be honest but would using them mean that each celery process could have its own orchestrator instance and if one of the orchestrators executed a critical section like `assign_worker` to change the states of `free_workers` and `busy_workers` another celery process with a different orchestrator instance would be locked and unable to access or modify its own `free_workers` and `busy_workers`? 

I am a bit unsure how much we would have to modify the orchestrator class to be able to have different instances whose `free_workers` and `busy_workers` remain correctly synchronized with one another or how complicated these changes would be which is why I suggested using one orchestrator instance that is accessible and modifiable via http endpoints

But if you have a good idea how to implement this with distributed locks we can definitely do that as well

LukaDeka replied:
Pipes for inter-process communication could also be used, but a lock also makes sense.

I don't have any experience with either, but I think we should close this pull request first before starting to work on that.

LukaDeka commented on embark/workers/tasks.py:
```diff
@@ -146,14 +167,141 @@ def update_worker_info():
+        logger.info("[Worker %s] Setting the worker as free.", worker.id)
+        return False
+    finally:
+        client.close()
```
good catch!

LukaDeka commented on embark/workers/tasks.py:
```diff
@@ -146,14 +167,141 @@ def update_worker_info():
+                # Fetch logs for the last time
+                _fetch_analysis_logs(worker)
+
+                orchestrator.release_worker(worker)
```
I'll add a `free_worker` method that just sets the worker as free. This can be refactored later.

LukaDeka commented:
I fixed the bugs in the latest commit and merged main.

These changes are untested, I plan to test it tomorrow.

**We should definitely focus on merging this issue as soon as possible** since over three big changes were implemented by me in this branch that didn't get merged from this branch, but from different branches that implemented basically the same refactorings. I noticed this while merging `main` for the 3rd time.


What still needs to be done:

* Ensure that `7z` is installed (checked) during the installation of the offline worker nodes / regular nodes / EMBArk
* Fix the `emba_run.log` issue (which in my opinion has nothing to do with this issue)
* Introduce inter-process communication between tasks/embark

LukaDeka commented:
I didn't add `apt-get autoremove --purge -y p7zip` to the uninstaller since `7z` is installed by default on the Ubuntu 22.04 server edition.

I added it to the installer to ensure the package is there.

ashiven commented on embark/workers/orchestrator.py:
```diff
@@ -144,10 +156,7 @@ def release_worker(self, worker: Worker):
-                del self.busy_workers[worker.ip_address]
-                worker.analysis_id = None
-                worker.save()
+                self.free_worker(worker)
```
You are probably going to run into a deadlock here because the same `lock` that was acquired a couple lines earlier, `free_worker` tries to acquire again and will be unable to acquire

LukaDeka replied:
Good catch!

LukaDeka commented:
Copy-pasting my previous comment about the `LogReader` because it's relevant:


After debugging, I discovered that emba_run.log, which is normally in {(analysis.path_to_logs).parent}/emba_run.log does NOT get created in the worker nodes.

I don't know where and if this file exists on the workers.

My suspicion is that stdout from the emba_run command is caught by EMBArk and not EMBA and is constantly being logged to the file.

Anyway, the absence of this file is exactly what causes this error in the menu, when the user tries to see live logs:
File is not yet available

This is why I believe correctly starting the analysis and making sure the logs are displayed properly should be part of a different feature and not of the "fetch logs" issue.

ClProsser commented on embark/workers/tasks.py:
```diff
@@ -143,14 +164,144 @@ def update_worker_info():
+
+    except Exception as exception:
+        logger.error("[Worker %s] Unexpected exception: %s", worker.id, exception)
+        logger.info("[Worker %s] Returning status free.", worker.id)
```
What is the purpose of this log message?

ClProsser commented on embark/workers/tasks.py:
```diff
@@ -174,7 +325,16 @@ def start_analysis(worker_id, emba_cmd: str, src_path: str, target_path: str):
     exec_blocking_ssh(client, f"sudo rm -rf {settings.WORKER_EMBA_LOGS}")
     exec_blocking_ssh(client, "sudo rm -rf ./terminal.log")
-    exec_blocking_ssh(client, f"sudo sh -c '{emba_cmd}' >./terminal.log 2>&1")
+    client.exec_command(f"sudo sh -c '{emba_cmd}' >./terminal.log 2>&1")  # nosec
```
This statement does not work if one uses a user != root and the user is not added to the sudoers file. I'll accept this anyway, as every user is added to the sudoers file soon.

ClProsser commented on embark/workers/update/deps_installer.sh:
```diff
@@ -35,6 +35,7 @@ apt-get install -y docker-ce docker-ce-cli containerd.io docker-buildx-plugin do
 
 apt-get install -y inotify-tools
 apt-get install -y libnotify-bin
+apt-get install -y p7zip 
```
The dependency can't be installed as we are on a *offline* worker. 
You have to download it on the EMBArk host. Thus, add the dependency to `embark/workers/update/deps_host.sh` using another call to `downloadPackage`

ClProsser commented on embark/workers/update/update.py:
Note that this change is not required, as the whole topic of putting the password into stdin is dropped using another issue. I just wanted to add that dumping lines which equal the password is not ideal (take a generic password like `root`, which might also be on the output with no relation to the password).

ClProsser reviewed (changes requested):
Please consider the latest feedback. Once this is fixed, I'll re-test it on my VM

LukaDeka commented on embark/workers/tasks.py:
```diff
@@ -107,14 +127,144 @@ def update_worker_info():
+        os.makedirs(f"{settings.MEDIA_ROOT}/log_zip/", exist_ok=True)
+
+        # Fetch zip file
+        sftp_client = client.open_sftp()
+        sftp_client.get("/root/emba_logs.zip", local_zip_path)
+        sftp_client.close()
+        client.close()
```
`settings.WORKER_EMBA_LOGS` is also hardcoded, as well as the firmware dir. I assumed we always use `root` as the user.

I added the environment variable and you suggested to remove it, so I did.

My plan in this issue is not to refactor everything and make sure all other issues work together (that aren't implemented yet), but rather to focus on the issue itself.

With the current configuration, `root` as the user is assumed since other paths are also hardcoded.

Thus I'm marking this as resolved.

ashiven commented on embark/workers/orchestrator.py:
```diff
@@ -126,6 +126,18 @@ def assign_worker(self, worker: Worker, task: OrchestratorTask):
         with self.lock:
             self._assign_worker(worker, task)
 
+    def free_worker(self, worker: Worker):
+        """
+        Marks a busy worker as free.
+
+        :param worker: The worker to be marked free
+        """
+        with self.lock:
+            del self.busy_workers[worker.ip_address]
+            self.free_workers[worker.ip_address] = worker
+            worker.analysis_id = None
+            worker.save()
```
If possible, please replace this function with `release_worker` as suggested and delete the definition for `release_worker` below.
The `release_worker` definition that assigns another task right after releasing a worker is obsolete after the merge of Issue 59. 

```suggestion
    def release_worker(self, worker: Worker):
        """
        Release a busy worker from its current task and mark it as free.

        :param worker: The worker to be released
        :raises ValueError: If the worker is not busy
        """
        with self.lock:
            if worker.ip_address not in self.busy_workers:
                raise ValueError(f"Worker with IP {worker.ip_address} is not busy.")

            del self.busy_workers[worker.ip_address]
            self.free_workers[worker.ip_address] = worker
            worker.analysis_id = None
            worker.save()
```

ashiven commented on embark/workers/tasks.py:
```diff
@@ -108,14 +128,144 @@ def update_worker_info():
+                _fetch_analysis_logs(worker)
+
+                # Only assign a new task after soft reset is done
+                orchestrator.free_worker(worker)
```
```suggestion
                orchestrator.release_worker(worker)
```

ashiven commented on embark/workers/tasks.py:
```diff
@@ -108,14 +128,144 @@ def update_worker_info():
+        except Exception as exception:
+            logger.error("[Worker %s] Unexpected exception: %s", worker.id, exception)
+            # TODO: Better handle exceptions
+            orchestrator.free_worker(worker)
```
```suggestion
            orchestrator.release_worker(worker)
```

ashiven reviewed (changes requested):
If you can, please delete the `free_worker` definition and replace it with the `release_worker` definition as suggested. We won't need the current `release_worker` definition any more after Issue 59 has been merged

LukaDeka commented:
I deleted `free_worker` and replaced it with `release_worker` as well as updated the unit tests that were failing.

I fixed the `root` issue by always placing the zip file in `client.ssh_user`'s homedir.

To move the file would be illogical since the zip file would not be updated, but rather generated each time. Thus big performance hit.

ashiven reviewed (approved):
Great work!

ClProsser commented on embark/workers/tasks.py:
```diff
@@ -108,14 +128,148 @@ def update_worker_info():
+    try:
+        client = worker.ssh_connect()
+
+        cmd = "sudo docker ps -q"
+        docker_output = exec_blocking_ssh(client, cmd)
```
Edit: Please retest this once this is fixed. This might not be sufficient.

LukaDeka replied:
My previous comment is a bit outdated.

Now, I make sure to wait 10 seconds before starting checking, if the docker container is running. That 10 seconds should be more than enough.

LukaDeka commented on embark/workers/tasks.py:
```diff
@@ -108,14 +128,148 @@ def update_worker_info():
+
+        # Fetch zip file
+        sftp_client = client.open_sftp()
+        sftp_client.get("/root/emba_logs.zip", local_zip_path)
```
I overlooked this, my bad

LukaDeka commented on embark/workers/tasks.py:
```diff
@@ -139,7 +293,16 @@ def start_analysis(worker_id, emba_cmd: str, src_path: str, target_path: str):
     exec_blocking_ssh(client, f"sudo rm -rf {settings.WORKER_EMBA_LOGS}")
     exec_blocking_ssh(client, "sudo rm -rf ./terminal.log")
-    exec_blocking_ssh(client, f"sudo sh -c '{emba_cmd}' >./terminal.log 2>&1")
+    client.exec_command(f"sudo sh -c '{emba_cmd}' >./terminal.log 2>&1")  # nosec
```
Thanks for the tip! I'll add this

LukaDeka commented on embark/workers/tasks.py:
```diff
@@ -108,14 +128,148 @@ def update_worker_info():
+            analysis = FirmwareAnalysis.objects.get(id=worker.analysis_id)
+            if analysis.status["finished"] or not is_running:
+                # Fetch logs for the last time
+                _fetch_analysis_logs(worker)
```
This is resolved by the refactor you suggested with the `while True` loop.

LukaDeka commented on embark/workers/tasks.py:
```diff
@@ -108,14 +128,148 @@ def update_worker_info():
+            logger.error("[Worker %s] Unexpected exception: %s", worker.id, exception)
+            # TODO: Better handle exceptions
+            orchestrator.release_worker(worker)
+            worker_soft_reset_task(worker, worker.configurations.first().id)
```
My mistake, I fixed the argument.

I reordered it correctly, and `release_worker` now has a lock so there should be no more issues here.

ClProsser reviewed (changes requested):
Now this looks great. I ran a full analysis as root and as a different user and it worked times. 

I suggested final comments (which are not functional, more about code quality). If you agree, please apply the suggested fixes. Once this is done, I'll merge it.

LukaDeka commented on embark/workers/update/deps_installer.sh:
```diff
@@ -35,6 +35,7 @@ apt-get install -y docker-ce docker-ce-cli containerd.io docker-buildx-plugin do
 
 apt-get install -y inotify-tools
 apt-get install -y libnotify-bin
+apt-get install -y p7zip-full
```
Valid point!

ClProsser commented on embark/workers/tasks.py:
```diff
@@ -138,8 +139,141 @@ def start_analysis(worker_id, emba_cmd: str, src_path: str, target_path: str):
+
+        except Exception as exception:
+            logger.error("[Worker %s] Unexpected exception: %s", worker.id, exception)
+            if worker in orchestrator.get_busy_workers().values():
```
Here you forgot the refactor. I'll add that in another issue, as I have to add sth anyway.
